src/controllers/auth-controller.ts

```diff
@@ -10,22 +13,97 @@ export default class AuthController extends AbstractController {
 
   private users: Repository<User>;
 
+  private tokens: Repository<RefreshToken>;
+
   public init(): Router {
     this.router = Router();
     this.users = getRepository(User);
+    this.tokens = getRepository(RefreshToken);
+    
     // @ts-ignore
     this.router.post('/signup', validateSchema(Registration), this.signUp);
+    this.router.post('/logout', this.logout);
+    this.router.post('/login', this.login);
+    this.router.post('/refresh', this.refresh);
 
+    /* use auth.required to secure route */
+    this.router.get('/test', auth.required, this.test)
     return this.router;
   }
 
   private signUp = async (req: Request, res: Response, next: NextFunction): Promise<void> => {
     try {
       const user = await User.create(req.body);
       await this.users.save(user);
-      res.json(user);
+      const { token, refreshToken } = singTokens({ userId: user.id, userRole: user.role });
+      await this.tokens.save(new RefreshToken(refreshToken));
+      res.json({ user, token: `Bearer ${token}`, refreshToken });
     } catch (e) {
       next(e);
     }
   };
+
+  private logout = async (req: Request, res: Response, next: NextFunction) => {
+    const { refreshToken }: { refreshToken: string } = req.body;
+    if (refreshToken) {
+      try {
+        const token = await this.tokens.findOne({ token: refreshToken });
+        if (token) {
+          await this.tokens.delete(token.id);
+          res.status(200).end();
+        }
+      } catch (e) {
+        next(e);
+      }
+    }
+    res.status(400).end();
+  };
+
+  private login = async (req: Request, res: Response, next: NextFunction) => {
+    const { email, password } = req.body;
+    if (!email || !password) {
+      res.status(403).end();
+    }
+    return passport.authenticate('local', { session: false }, async (error: {}, user: User) => {
+      if (error) {
+        res.status(403).json({ error });
+      } else {
+        try {
+          const { token, refreshToken } = singTokens({ userId: user.id, userRole: user.role });
+          await this.tokens.save(new RefreshToken(refreshToken));
+          res.json({ user, token: `Bearer ${token}`, refreshToken });
+        } catch (e) {
+          next(e);
+        }
+      }
+    })(req, res, next);
+  };
+
+  private refresh = async (req: Request, res: Response) => {
```

Comment:
This endpoint needs to be reworked as using unoptimal flow and unappropriate error statuses. At least it also have **critical bug**.

close request earlier, before retrieving token from db, if there isn't token in body -- just return 400.

then, find this token in DB, if it not exists, this means that it was stolen with huge confidence, please, **log** such cases. Here can be passed as 401, as 403.

`verifyRefreshToken` and `users.findOne` can be done in parallel for bust up.

If there not such user, also you should make early return with 403. It's exact case, when somebody use token of deleted user, that's we already had discussed with you. Needs **additional** logging.

Now we near founded by me **bug**: you remove token only in catch. It causes that all previous tokens will be valid and creates security issue, also it comes about gradual overflow DB with invalid tokens. And just want you remind that each call to `/refresh` erase previous token, not only if it was expired.

All this is achivable with one top level `try..catch`.

Sure that there are other issues to fix, consider this as initial request for this endpoint)

Reply:
yep, you should split up the logic and process all possible cases with error handling